test/crypto_sign/functest.c

```diff
@@ -124,8 +134,25 @@ static int test_sign(void) {
         RETURNS_ZERO(crypto_sign_keypair(pk + 8, sk + 8));
 
         randombytes(m + 8, MLEN);
+        // Fill the sm buffer with random bytes
+        randombytes(sm_random_cmp, MLEN + CRYPTO_BYTES);
+        memcpy(sm + 8, sm_random_cmp, MLEN + CRYPTO_BYTES);
+
+        /*
+         * With this buffer marked as undefined, valgrind will detect
+         * cases where the signing code depends on the value of the tail
+         * of the buffer.
+         */
+        VALGRIND_MAKE_MEM_UNDEFINED(sm + 8, MLEN + CRYPTO_BYTES);
```

Comment:
It's a bit of a shame that Valgrind is required, while I have the impression that `asan` is a bit more mature nowadays (and faster). We'd ideally also want to trigger these in ASAN. Maybe we can kill two birds with one stone by instead using `malloc` to request fresh memory inside the loop. This should always be unallocated, and perhaps a bit easier for both Valgrind and ASAN to measure than the stack.

What do you think?

Reply:
But of course, then we can't do both checks in one go that we are doing here...

Reply:
For that I believe we'd have to pass `-fsanitize=memory`, which is supported by Clang but not GCC. As far as I can tell, ASAN and UBSAN don't catch uninitialized reads. I suppose the sanitizer tests could be forced to use Clang, but apparently `-fsanitize=memory` and `-fsanitize=address` are incompatible, so it would require two passes.